### PR TITLE
make wallet sortable

### PR DESCRIPTION
**Pull request by brootle**
sorting Wallet by Username, Coin Price and by USD Value

added short video demo

GIF-c16062b8ac230e390c807f49e9be2a3c

### CODE REVIEW

**Comment by tijno** [comment]
i bet you resubmitted just so you could have #234 :)

ive been keen to have sorting for a while. Maybe at a little gif to the description to show sorting in action?

**Comment by brootle** [comment]
> i bet you resubmitted just so you could have #234 :)
> 
> ive been keen to have sorting for a while. Maybe at a little gif to the description to show sorting in action?

Epic updates need epic numbers! 😉

File: src/app/wallet/wallet.component.ts [hunk 1]
```diff
@@ -14,6 +14,10 @@ export class WalletComponent implements OnInit {
   hasUnminedCreatorCoins: boolean;
   showTransferredCoins: boolean = false;
 
+  sortedFromHighToLow: number = 0;
```
Comment (lazynina) [inline src/app/wallet/wallet.component.ts:17]: give this variable a more logical name. perhaps `sortedUSDValueFromHighToLow`?

File: src/app/wallet/wallet.component.ts [hunk 2]
```diff
@@ -55,6 +59,106 @@ export class WalletComponent implements OnInit {
...
+      this.sortHodlingsUsernameToHigh(this.usersYouReceived);      
+      this.sortedUsernameFromHighToLow = 0;
+    } else {
+      this.sortHodlingsUsernameToLow(this.usersYouPurchased);
+      this.sortHodlingsUsernameToLow(this.usersYouReceived);  
+      this.sortedUsernameFromHighToLow = 1;    
+    }
+    this.sortedFromHighToLow = - 1;
+    this.sortedPriceFromHighToLow = -1;  
+  }
```
Comment (lazynina) [inline src/app/wallet/wallet.component.ts:157]: seems like this code is a bit duplicative. Anyway we can refactor such that there is one function that handles the sorting of the wallet?

File: src/app/wallet/wallet.component.ts [hunk 3]
```diff
@@ -14,6 +14,10 @@ export class WalletComponent implements OnInit {
   hasUnminedCreatorCoins: boolean;
   showTransferredCoins: boolean = false;
 
+  sortedFromHighToLow: number = 0;
+  sortedPriceFromHighToLow: number = -1;
```
Comment (lazynina) [inline src/app/wallet/wallet.component.ts:18]: I think 1 should signify sort ascending, -1 should signify sort descending, and 0 means not sorting on this column.  It's a bit more intuitive that way.

File: src/app/wallet/wallet.component.ts [hunk 4]
```diff
@@ -55,6 +59,106 @@ export class WalletComponent implements OnInit {
...
+
+    hodlings.sort((a: BalanceEntryResponse, b: BalanceEntryResponse) => {
+      return(
+        a.ProfileEntryResponse.Username.localeCompare(b.ProfileEntryResponse.Username)
+      )
+    });
+
+  }  
+
+  sortHodlingsUsernameToLow(hodlings: BalanceEntryResponse[]): void {
```
Comment (lazynina) [inline src/app/wallet/wallet.component.ts:135]: maybe we just add a param to one of these to indicate that we reverse it.  These are basically the same functions. e.g. `sortHodlingsUsername(hodlings: BalanceEntryResponse, descending: boolean): void`
and if descending, just multiply by -1

Comment (lazynina) [reply src/app/wallet/wallet.component.ts:135]: apply the same logic to the above functions for price and USD value

**Review by lazynina — CHANGES_REQUESTED** [review]
Some nits about duplicated code. Also, can we remove all the spacing changes in the html from this PR? I don't think it conforms with the linting rules defined in the repo -- still want to get CI running linting.

File: src/app/wallet/wallet.component.ts (hunk 4 above)
Comment (brootle) [reply src/app/wallet/wallet.component.ts:135]: Something like this?

```
  sortHodlingsUsername(hodlings: BalanceEntryResponse[], descending: boolean): void{
    if (descending){
      hodlings.sort((a: BalanceEntryResponse, b: BalanceEntryResponse) => {
        return (
          a.ProfileEntryResponse.Username.localeCompare(b.ProfileEntryResponse.Username)
        )
      });
    }else {
      hodlings.sort((a: BalanceEntryResponse, b: BalanceEntryResponse) => {
        return (
          b.ProfileEntryResponse.Username.localeCompare(a.ProfileEntryResponse.Username)
        )
      });
    }
    this.sortedUsernameFromHighToLow = descending ? -1 : 1;
    this.sortedUSDValueFromHighToLow = 0;
    this.sortedPriceFromHighToLow = 0;
  }

  toggleSortUsername(){
    if (this.sortedUsernameFromHighToLow === 1 || this.sortedUsernameFromHighToLow === 0){     
      this.sortHodlingsUsername(this.usersYouPurchased, true);
      this.sortHodlingsUsername(this.usersYouReceived, true);
    } else {
      this.sortHodlingsUsername(this.usersYouPurchased, false);
      this.sortHodlingsUsername(this.usersYouReceived, false);   
    } 
  }
```

Comment (lazynina) [reply src/app/wallet/wallet.component.ts:135]:
```
  sortHodlingsUsername(hodlings: BalanceEntryResponse[], descending: boolean): void {
    this.sortedUsernameFromHighToLow = descending ? -1 : 1;
    this.sortedUSDValueFromHighToLow = 0;
    this.sortedPriceFromHighToLow = 0;
    hodlings.sort((a: BalanceEntryResponse, b: BalanceEntryResponse) => {
        return (
          this.sortedUsernameFromHighToLow * b.ProfileEntryResponse.Username.localeCompare(a.ProfileEntryResponse.Username)
        )
      });
  }

  toggleSortUsername(){
    if (this.sortedUsernameFromHighToLow === 1 || this.sortedUsernameFromHighToLow === 0){     
      this.sortHodlingsUsername(this.usersYouPurchased, true);
      this.sortHodlingsUsername(this.usersYouReceived, true);
    } else {
      this.sortHodlingsUsername(this.usersYouPurchased, false);
      this.sortHodlingsUsername(this.usersYouReceived, false);   
    } 
  }
```

Comment (lazynina) [reply src/app/wallet/wallet.component.ts:135]: I was thinking something like that so we don't have to have two different calls to sort

Comment (brootle) [reply src/app/wallet/wallet.component.ts]: ok, refactored, have a look

**Review by lazynina — APPROVED** [review]

**Comment by lazynina** [comment]
Great work! Thanks @brootle

**Pull request merged**